op-challenger/game/fault/contracts/multicall.go

```diff
@@ -0,0 +1,170 @@
+package contracts
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"math/big"
+
+	"github.com/ethereum-optimism/optimism/op-service/sources"
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/rpc"
+)
+
+// Note: All of this stuff would wind up moving to somewhere in op-service so it can be easily reused.
+
+type EthRpc interface {
+	CallContext(ctx context.Context, out interface{}, method string, args ...interface{}) error
+	BatchCallContext(ctx context.Context, b []rpc.BatchElem) error
+}
+
+type ContractCall struct {
+	Abi    *abi.ABI
+	Addr   common.Address
+	Method string
+	Args   []interface{}
+}
+
+func NewContractCall(abi *abi.ABI, addr common.Address, method string, args ...interface{}) *ContractCall {
```

Comment:
Go bindings wrap this into a bound contract & then you can create a specific call by only passing the methods & args. I kinda like it, but also have both or this isn't bad.

Reply:
I'd say the aim of `NewContractCall` is to be the low level building block that lets you manually build the wrapper that provides a nice API (which is the `FaultDisputeGameContract` in this example). It's mostly nice to not have to repeat the rather long winded and very error-prone interactions with the ABI to pack args and `ContractCall` handles that for you (in a way that then works with making multiple calls in a single batch request)

Reply:
Talking in code, I'm suggesting the following as syntatic sugar/convenience

```
type BoundContract struct {
    Abi  *abi.ABI
    Addr common.Address
}

func (b *BoundContract) NewContractCall(method string, args ...interface{}) *ContractCall {
    return &ContractCall{
		Abi:    b.abi,
		Addr:   b.addr,
		Method: method,
		Args:   args,
	}
}


```
It's not a huge deal, but I think it makes it slightly nicer in some cases.

Reply:
oh sorry, yeah I completely misunderstood what you were saying. :)

Yes that's a great idea. I suspect we'd actually always use the `BoundContract` type approach since you'd create `BoundContract` when creating the wrapper (eg `FaultDisputeGameContract`) and would then naturally use it everywhere. Both options would pretty naturally be supported though I think.